File: rag/src/context-pack.js
/**
 * context-pack.js — Context Pack 构建器
 *
 * 职责：
 * - 将检索结果组装成 Agent 可直接消费的 ContextPack
 * - 按任务类型组织知识结构
 * - 优化呈现顺序：规则优先、高优先级优先
 */

/**
 * 构建 Context Pack
 * @param {string} query - 原始查询
 * @param {string} taskType - 路由后的任务类型
 * @param {object} filterInfo - 使用的过滤条件
 * @param {Array} results - 检索结果
 * @param {object} meta - 检索元信息
 * @returns {object} ContextPack
 */
export function buildContextPack(query, taskType, filterInfo, results, meta) {
  // 按 source_type 分组排序: rule > reference > project
  const sorted = [...results].sort((a, b) => {
    const typeOrder = { rule: 0, reference: 1, project: 2 };
    const ta = typeOrder[a.source_type] || 3;
    const tb = typeOrder[b.source_type] || 3;
    if (ta !== tb) return ta - tb;
    return b.score - a.score;
  });

  // 格式化每个结果
  const formattedResults = sorted.map(r => ({
    chunk_id: r.chunk_id,
    title: r.title,
    score: r.score,
    reason: r.reason,
    snippet: r.snippet,
    source_path: r.source_path,
    source_type: r.source_type,
    category: r.category,
    tags: r.tags || []
  }));

  // 按 source_type 总结知识分布
  const sourceBreakdown = {};
  for (const r of formattedResults) {
    const st = r.source_type;
    if (!sourceBreakdown[st]) sourceBreakdown[st] = 0;
    sourceBreakdown[st]++;
  }

  // 生成简短的知识概要
  const knowledgeSummary = generateSummary(query, taskType, formattedResults);

  return {
    task_type: taskType,
    query: query,
    filters: filterInfo || {},
    results: formattedResults,
    knowledge_summary: knowledgeSummary,
    source_breakdown: sourceBreakdown,
    meta: meta || {}
  };
}

/**
 * 生成简短的知识概要摘要
 */
function generateSummary(query, taskType, results) {
  if (results.length === 0) {
    return '未检索到相关知识。';
  }

  const topSources = results.slice(0, 3).map(r => `"${r.title}"`);

  const stageMap = {
    humanization: '语言自然化和去AI味',
    outline_review: '大纲评估和结构检查',
    chapter_prewrite: '写前准备和章节规划',
    ideation: '构思和灵感激发',
    consistency_check: '一致性和逻辑检查',
    rhythm_review: '阅读体验和节奏优化',
    genre_routing: '题材选型和路线确定'
  };

  const stageHint = stageMap[taskType] || taskType;

  return `查询「${query}」匹配到 ${results.length} 条知识，类型为「${stageHint}」。
其中知识来源：${topSources.join('、')}。`;
}

/**
 * 生成纯文本版的 context pack（供 Agent prompt 直接注入）
 */
export function contextPackToText(contextPack) {
  const lines = [];
  lines.push(`## 知识上下文包`);
  lines.push(`任务类型：${contextPack.task_type}`);
  lines.push(`查询：${contextPack.query}`);
  lines.push(`知识来源分布：${JSON.stringify(contextPack.source_breakdown)}`);
  lines.push(``);

  if (contextPack.results.length === 0) {
    lines.push(`> ⚠️ 未检索到相关知识。`);
    lines.push(``);
    return lines.join('\n');
  }

  lines.push(`### 相关知识`);

  for (let i = 0; i < contextPack.results.length; i++) {
    const r = contextPack.results[i];
    lines.push(``);
    lines.push(`**${i + 1}. ${r.title}**  (相关度: ${r.score})`);
    lines.push(`- 来源: \`${r.source_path}\``);
    lines.push(`- 类型: ${r.source_type}`);
    if (r.tags && r.tags.length > 0) {
      lines.push(`- 标签: ${r.tags.join(', ')}`);
    }
    lines.push(`- 检索依据: ${r.reason}`);
    lines.push(``);
    lines.push(`> ${r.snippet}`);
    lines.push(``);
    lines.push(`---`);
  }

  return lines.join('\n');
}

export default { buildContextPack, contextPackToText };

File: rag/src/context_pack.py
"""
context_pack.py - 根据检索结果构建可注入提示词的上下文包。
"""


def build_context_pack(query, retrieval_result, task_type=None, project_hint=None):
    """构建标准化 context pack。"""
    results = retrieval_result.get("results", [])
    meta = retrieval_result.get("meta", {})

    type_order = {"rule": 0, "knowledge": 1, "reference": 2, "project": 3}
    sorted_results = sorted(results, key=lambda r: type_order.get(r.get("source_type", ""), 9))

    top_sources = [r.get("title", "") for r in sorted_results[:5]]
    knowledge_summary = f"检索到 {len(sorted_results)} 条相关知识点"
    if top_sources:
        knowledge_summary += "，涉及: " + "、".join(top_sources)

    source_breakdown = {"rule": 0, "knowledge": 0, "reference": 0, "project": 0}
    for result in sorted_results:
        source_type = result.get("source_type", "")
        if source_type in source_breakdown:
            source_breakdown[source_type] += 1

    return {
        "task_type": task_type or meta.get("task_type", "general"),
        "query": query,
        "project_hint": project_hint or "",
        "filters": {
            "categories": [],
            "stages": [],
        },
        "results": [
            {
                "chunk_id": r["chunk_id"],
                "title": r["title"],
                "score": r["score"],
                "reason": r["reason"],
                "snippet": r["snippet"],
                "source_path": r["source_path"],
                "source_type": r["source_type"],
                "category": r.get("category", ""),
                "tags": r.get("tags", []),
            }
            for r in sorted_results
        ],
        "knowledge_summary": knowledge_summary,
        "source_breakdown": source_breakdown,
        "meta": {
            "total_candidates": meta.get("total_candidates", 0),
            "elapsed_ms": meta.get("elapsed_ms", 0),
            "task_type": meta.get("task_type", ""),
            "confidence": meta.get("confidence", 0),
        },
    }


def context_pack_to_text(context_pack):
    """把 context pack 格式化成纯文本，供 Agent prompt 注入。"""
    lines = []
    lines.append("---")
    lines.append("[知识上下文]")
    lines.append(f"任务类型: {context_pack.get('task_type', 'general')}")
    lines.append(f"查询: {context_pack.get('query', '')}")

    source_breakdown = context_pack.get("source_breakdown", {})
    names = {
        "rule": "规则",
        "knowledge": "知识包",
        "reference": "参考",
        "project": "项目",
    }
    parts = []
    for source_type in ("rule", "knowledge", "reference", "project"):
        count = source_breakdown.get(source_type, 0)
        if count > 0:
            parts.append(f"{names.get(source_type, source_type)} x {count}")
    lines.append(f"来源: {', '.join(parts) if parts else '无'}")

    if context_pack.get("knowledge_summary"):
        lines.append(f"摘要: {context_pack['knowledge_summary']}")

    lines.append("")

    for index, result in enumerate(context_pack.get("results", []), 1):
        source_type = result.get("source_type", "")
        source_type_label = names.get(source_type, source_type)
        lines.append(
            f"{index}. [{source_type_label}] {result.get('title', '')} "
            f"(相关性: {result.get('score', 0)})"
        )
        lines.append(f"   {result.get('reason', '')}")
        lines.append(f"   {result.get('snippet', '')}")
        lines.append("")

    lines.append("---")
    return "\n".join(lines)
